cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,530 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <thrust/execution_policy.h>
+#include <thrust/for_each.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/tuple.h>
+#include <thrust/device_vector.h>
+#include <thrust/merge.h>
+
+#include <algorithm>
+#include <utility>
+#include <vector>
+#include <memory>
+#include <type_traits>
+
+#include <nvstrings/NVCategory.h>
+
+#include <cudf/cudf.h>
+#include <cudf/types.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <rmm/thrust_rmm_allocator.h>
+#include <utilities/cuda_utils.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <cudf/null_mask.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+
+#include <cudf/merge.hpp>
+
+namespace { // anonym.
+using index_type = thrust::tuple<side, cudf::size_type>; // `thrust::get<0>` indicates left/right side, `thrust::get<1>` indicates the row index
+
+
+/**
+ * @brief Extractor is the primary template for "tagged-index" data extraction (side and index).
+ *
+ * It serves as an interface to adhere to and a uniform way of accessing the "tagged-index" abstraction. 
+ * Hence, it is purposely implementation-free 
+ * because different tagged index types may have different representations, getters, etc.;
+ * e.g., index_type::get_side/get_index;
+ * or
+ * get<0>(tuple<side, size_t>), get<1>(tuple<side, size_t>).
+ *
+ * This class is meant to to be fully specialized for different "tagged-index" classes.
+ */
+template<typename TaggedIndexT>
+struct Extractor
```

Comment:
The reason I put it in place was in case someone wanted to reuse this code, but with a different abstraction for 'TaggedIndexT`. So, yes, purpose was to make it more generic. I will remove it.

Reply:
Removed / simplified.